### PR TITLE
[hybrid enhance] add flag to control the avg position for grad merge under pipeline mode

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
New features

### PR changes
Others

### Describe
Support pure dp kind 'average for gradient merge after merging' operation for pp mode.
Add a new flag inside gradient_scale_configs: `scale_gradient`, which default value is False, to control where to do the average operation for gradient merge.
If `gradient_scale_configs['scale_gradient'] = False`, the pipeline will scale the loss@Grad (divide by number of micro steps) for each micro step, then do the gradient merge. It does the gradient merge avg op before sum.
Otherwise, if `gradient_scale_configs['scale_gradient'] = True`, the pipeline will do the gradient merge firstly. Then, during the optimize stage, the merged gradient will be divided by the number of micro steps. It does the gradient merge avg op after sum.

Hot to use:
```python
strategy = paddle.distributed.fleet.DistributedStrategy()
strategy.gradient_scale_configs = {
    'scale_strategy': 'avg',
    'scale_gradient': True
}
```

Loss compared under fp16_allreduce
image

### CODE REVIEW

File: paddle/fluid/framework/distributed_strategy.proto [hunk 1]
```diff
@@ -162,6 +162,7 @@ message PipelineConfig {
   optional int32 accumulate_steps = 2 [ default = 1 ];
   optional string schedule_mode = 3 [ default = '1F1B' ];
   optional bool p2p_cache_shape = 4 [ default = true ];
+  optional bool grad_merge_avg_after_sum = 5 [ default = false ];
```
Comment (wangxicoding) [inline paddle/fluid/framework/distributed_strategy.proto]: 放到专门的loss scale那个选项里吧

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 2]
```diff
@@ -362,7 +368,11 @@ def _insert_allreduce_for_pp(self, params_grads):
             main_block, strategy=strategy, shard=shard)
 
         len_of_ops = len(main_block.ops)
-        first_optimize_op_index = get_first_optimize_op_idx(main_block)
+        if self.grad_merge_avg_after_sum:
+            first_optimize_op_index = self._avg_grad_merge_after_sum(
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: 验证下和fp16_allreduce一起时的精度

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 3]
```diff
@@ -429,6 +439,27 @@ def _insert_allreduce_for_pp(self, params_grads):
 
         # FIXME(wangxi): if fp16_allreduce, put cast fp16->fp32 to there?
 
+    def _avg_grad_merge_after_sum(self, main_block, accumulated_grad_names):
+        # For pp, do the avg operation for gradient merge after merging
+        # the gradient to meet the logic for gradient merge under pure dp.
+        tmp_first_opt_idx = get_first_optimize_op_idx(main_block)
+        for grad in accumulated_grad_names:
+            main_block._insert_op_without_sync(
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: 开启amp的时候，可以对unscale进行优化，可以问下华超

File: paddle/fluid/framework/distributed_strategy.proto (hunk 1 above)
Comment (FeixLiu) [reply paddle/fluid/framework/distributed_strategy.proto]: done

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py (hunk 2 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: will do

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py (hunk 3 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: finish

File: paddle/fluid/framework/distributed_strategy.proto [hunk 4]
```diff
@@ -133,6 +133,10 @@ message GradientScaleConfig {
   // Else if sum, the gradient will accumulated among multiple
   // devices.
   optional string scale_strategy = 1 [ default = 'avg' ];
+  // The avg_loss flag is used to determine the position of average
+  // If avg_loss is True, it will avg the loss before grad merge.
+  // Otherwise, it will do grad merge firstly, then avg the grad after merging.
+  optional bool avg_loss = 2 [ default = true ];
```
Comment (wangxicoding) [inline paddle/fluid/framework/distributed_strategy.proto]: 这个名字感觉有点困惑，实际上是对loss@grad操作的。是不是改成`scale_gradient`或者`avg_gradient`好一些，默认为fasle。
另外这个应该不止针对grad merge，普通的dp之类也需要这样的策略，这个以后再修改

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 5]
```diff
@@ -429,6 +442,52 @@ def _insert_allreduce_for_pp(self, params_grads):
 
         # FIXME(wangxi): if fp16_allreduce, put cast fp16->fp32 to there?
 
+    def _avg_grad_merge_after_sum(self, main_block, accumulated_grad_names):
+        tmp_first_opt_idx = get_first_optimize_op_idx(main_block)
+        if self.user_defined_strategy.amp:
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: 这个地方，开启amp时，不一定需要开启动态loss scale，可能没有check_finite_and_unscale op。不过未来需要把amp里的loss scale策略统一到GradientScaleConfig里的

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 6]
```diff
@@ -429,6 +442,52 @@ def _insert_allreduce_for_pp(self, params_grads):
...
+            # For AMP, the avg operation can be simple done by modify the
+            # LossScaling op.
+            for idx, op in enumerate(main_block.ops):
+                if op.type == 'check_finite_and_unscale':
+                    loss_scale_name = op.input('Scale')[0]
+                    loss_scaling_var = main_block.var(loss_scale_name)
+                    loss_scale_tmp_var = main_block.create_var(
+                        name="loss_scale_tmp_var",
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: 这个name在原loss_scale_name加后缀吧

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 7]
```diff
@@ -429,6 +442,52 @@ def _insert_allreduce_for_pp(self, params_grads):
...
+                        shape=loss_scaling_var.shape,
+                        dtype=loss_scaling_var.dtype)
+                    main_block._insert_op(
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: _insert_op_without_sync 加点速

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 8]
```diff
@@ -429,6 +442,52 @@ def _insert_allreduce_for_pp(self, params_grads):
...
+                        idx,
+                        type='scale',
+                        inputs={'X': loss_scaling_var},
+                        outputs={'Out': loss_scale_tmp_var},
+                        attrs={
+                            'scale': self._gradient_merge_acc_step,
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py:463]: 这个之后可能需要把dp的nranks给乘上去

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 9]
```diff
@@ -429,6 +442,52 @@ def _insert_allreduce_for_pp(self, params_grads):
...
+                    outputs={'Out': grad},
+                    attrs={
+                        'scale': 1.0 / self._gradient_merge_acc_step,
+                        'bias': 0.0,
+                        'bias_after_scale': False,
+                        OP_ROLE_KEY: OpRole.Optimize
+                    })
+            # Note(Yuang Liu): Need to move the first opt idx a little bit after
+            # to make sure the c_allreudce_sum ops are inserted behind scale ops
+            tmp_first_opt_idx += len(accumulated_grad_names)
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: 得先c_allreduce_sum再scale梯度吧

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py (hunk 8 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py:463]: will do later

File: paddle/fluid/framework/distributed_strategy.proto (hunk 4 above)
Comment (FeixLiu) [reply paddle/fluid/framework/distributed_strategy.proto]: done

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py (hunk 9 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py]: done

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 10]
```diff
@@ -429,6 +440,55 @@ def _insert_allreduce_for_pp(self, params_grads):
...
+                            OP_ROLE_KEY: OpRole.Optimize
+                        })
+                    op._rename_input(loss_scale_name, loss_scale_tmp_var_name)
+                    break
+        else:
+            # For pp, do the avg operation for gradient merge after merging
+            # the gradient to meet the logic for gradient merge under pure dp.
+            tmp_first_opt_idx = None
+            for idx, op in enumerate(main_block.ops):
+                if is_optimizer_op(op) and op.type != 'c_sync_comm_stream':
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py:475]: 这个插在c_sync_comm_stream前也没有关系

**Review by wangxicoding — APPROVED** [review]
LGTM
